### PR TITLE
XIVY-15533 Add ConditionBuilder

### PR DESCRIPTION
**Pull request by ivy-edp**
Here is my suggestion for the Condition Builder in UI Components (currently just a draft, as I still want to add tests, but you can take a look at it in the meantime):
In the Browser story, I demonstrate how to integrate the Condition Builder as a browser into the BrowserView, and in the ConditionBuilder story, I show how to use the condition builder independently (so it can already be used in the inscription as well).

### CODE REVIEW

File: packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  'is false': string;
+  'is empty': string;
+  'is not empty': string;
+  'less than': string;
+  'greater than': string;
+  'less or equal to': string;
+  'greater or equal to': string;
+};
+
+export type ConditionType = keyof TypeOperator;
```
Comment (ivy-edp) [inline packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx:20]: Unfortunately, I haven't found a way to define a type that takes all the values from, for example, typesOptions (since the typesOptions object can't be accessed at the time the type is generated). So i just use the keyof TypeOperator. Also not quiet sure about the naming of TypeOperator/ConditionType and LogicOperator/LogicalOperator

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    case 'is true':
+      return `${formattedArg1}${logicalOp}`;
+    case 'is false':
+      return `!${formattedArg1}${logicalOp}`;
+    case 'is empty':
+      return `${formattedArg1}.isEmpty()${logicalOp}`;
+    case 'is not empty':
+      return `!${formattedArg1}.isEmpty()${logicalOp}`;
+    default:
+      return `${formattedArg1} ${TypeOptions[con.operator]} ${formattedArg2}${logicalOp}`;
```
Comment (ivy-edp) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:108]: Here you can see that as a consequence that ConditionType directly uses keyof TypeOperator and not the actual value of the attribute, I need to search for the correct value in the implementation (if it's unclear, we can discuss it in person)

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  'less or equal to': '<=',
+  'greater or equal to': '>='
+};
+
+export const LogicalOperatorOptions: LogicOperator = {
+  'and': '&&',
+  'or': '||'
+};
+
+export const GenerateConditionString = (conditionMode: ConditionMode, conditionGroups: ConditionGroup[]) => {
```
Comment (ivy-edp) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:62]: I'm not sure if it's a good idea to define these const (TypeOptions, LogicalOperatorOptions, GeneratedConditionString) as exports. I did it because this allows me to easily reuse them in the Browser story.

File: packages/components/src/components/editor/browser/browser.stories.tsx [hunk 4]
```diff
@@ -37,6 +43,16 @@ const DefaultBrowser = ({ applyFn, applyBtn, initSearch }: DefaultBrowserProps)
...
 
+  const argumentInput = (value: string, onChange: (change: string) => void) => {
+    return <DialogBrowserInput value={value} onChange={onChange} />;
+  };
+  const condition = useConditionBuilder({
+    generateConditionString: GenerateConditionString,
+    typeOptions: TypeOptions,
+    logicalOperatorOptions: LogicalOperatorOptions,
+    argumentInput
+  });
```
Comment (ivy-lli) [inline packages/components/src/components/editor/browser/browser.stories.tsx:54]: I think I would directly include the function instead of define it above

```suggestion
  const argumentInput = (value: string, onChange: (change: string) => void) => {
    return <DialogBrowserInput value={value} onChange={onChange} />;
  };
  const condition = useConditionBuilder({
    generateConditionString: GenerateConditionString,
    typeOptions: TypeOptions,
    logicalOperatorOptions: LogicalOperatorOptions,
    argumentInput: (value, onChange) => <DialogBrowserInput value={value} onChange={onChange} />
  });
```

File: packages/components/src/components/editor/browser/browser.stories.tsx [hunk 5]
```diff
@@ -121,29 +138,33 @@ export const Default: Story = {
...
+  return (
+    <Dialog open={open} onOpenChange={setOpen}>
+      <InputGroup style={{ flex: '1' }}>
+        <Input value={value} onChange={e => onChange(e.target.value)} />
+        <DialogTrigger asChild>
+          <Button icon={IvyIcons.ListSearch} aria-label='Browser' />
+        </DialogTrigger>
+      </InputGroup>
+      <DialogContent style={{ height: '80vh' }}>
+        <DefaultBrowser
```
Comment (ivy-lli) [inline packages/components/src/components/editor/browser/browser.stories.tsx:152]: Isn't it possible to open in this browser also a condition builder? 🙈

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+          style={{ border: 'var(--basic-border)', maxHeight: '80px', padding: 'var(--input-padding)', borderRadius: 'var(--border-r2)' }}
+        >
+          {condition}
+        </pre>
+      </>
+    );
+  }
+};
+
+export const TypeOptions: TypeOperator = {
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:44]: this is an object and should not start uppercase
```suggestion
export const typeOptions: TypeOperator = {
```

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+  'is false': 'isFalse',
+  'is empty': 'isEmpty',
+  'is not empty': 'isNotEmpty',
+  'less than': '<',
+  'greater than': '>',
+  'less or equal to': '<=',
+  'greater or equal to': '>='
+};
+
+export const LogicalOperatorOptions: LogicOperator = {
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:57]: this is an object and should not start uppercase
```suggestion
export const logicalOperatorOptions: LogicOperator = {
```

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx (hunk 3 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:62]: I think this is fine. 
This stories are not build in the library bundle output. They only live in the storybook output.

Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:62]: But this is a function and should not be uppercase
```suggestion
export const generateConditionString = (conditionMode: ConditionMode, conditionGroups: ConditionGroup[]) => {
```

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+  return groupStrings.join('');
+};
+
+const getLogicalOperator = (index: number, totalConditions: number, logicalOperator: LogicalOperator) => {
+  return index < totalConditions - 1 ? ` ${LogicalOperatorOptions[logicalOperator].toLowerCase()} ` : '';
+};
+
+const formatCondition = (con: Condition, formattedArg1: string, formattedArg2: string, logicalOp: string) => {
+  switch (con.operator) {
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:98]: con is not a very good name IMO. I would call it condition

File: packages/components/src/components/editor/condition-builder/Condition.tsx [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+import { BasicSelect } from '@/components/common/select/select';
+import {
+  useConditionBuilderContext,
+  type ConditionType,
+  type LogicalOperator
+} from '@/components/editor/condition-builder/ConditionBuilderContext';
+import { IvyIcons } from '@axonivy/ui-icons';
+import { Label } from '@radix-ui/react-label';
+
+export interface Condition {
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/Condition.tsx:13]: I think it is not idea that the type here has the same name as the Component.

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx (hunk 2 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:108]: I'm not sure what you mean, but this is fine or not?
the con.operator is one of the keys of the TypeOperator, and the value for this key is the thing which is different and the thing we need to put here.

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+interface ConditionGroupProps extends Pick<ConditionProps, 'groupIndex'> {
+  group: ConditionGroup;
+  groupCount: number;
+}
+
+export const ConditionGroup = ({ group, groupIndex, groupCount }: ConditionGroupProps) => {
+  const { updateLogicalOperator, addCondition, removeConditionGroup, conditionMode, logicalOperatorOptions } = useConditionBuilderContext();
+
+  return (
+    <Flex direction='column' gap={2} className='condition-builder__group'>
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:23]: I assume you have this class here also for the impls so they can test this more easily?
I started here to streamline this additional class names a bit (see the other components). 
They always start with `ui-` it think, so in this case I would name them something like `ui-condition-builder-group`

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      <Flex
+        direction='column'
+        style={
+          conditionMode === 'nested-condition'
+            ? { border: 'var(--basic-border', borderRadius: 'var(--border-r2)', padding: 'var(--size-2)' }
+            : undefined
+        }
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:30]: I think it is sometimes fine to use inline styles in the impls, but here I think we should do all the same way, via a separate CSS file, and use the known variables for this variables :)
See the other components for how.

And stuff like the condition mode should be a simple data argument, so you can adjust the styles in the CSS file according to it:
`data-condition-mode={conditionMode}`

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+          />
+        ))}
+        <Button onClick={() => addCondition(groupIndex)} icon={IvyIcons.Plus} aria-label='Add Condition' variant='outline'>
+          Add Condition
+        </Button>
+      </Flex>
+      {groupIndex < groupCount - 1 && conditionMode === 'nested-condition' && (
+        <BasicSelect
+          items={Object.entries(logicalOperatorOptions)
+            .slice(0, -1)
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:55]: remove empty one?

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+            conditionsCount={group.conditions.length}
+          />
+        ))}
+        <Button onClick={() => addCondition(groupIndex)} icon={IvyIcons.Plus} aria-label='Add Condition' variant='outline'>
+          Add Condition
+        </Button>
+      </Flex>
+      {groupIndex < groupCount - 1 && conditionMode === 'nested-condition' && (
+        <BasicSelect
+          items={Object.entries(logicalOperatorOptions)
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:54]: this options to items mapping is very global, and should be only once IMO, so not on every rerender.
E.g. in a useMemo at the start of the component.

File: packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+type ConditionBuilderHookProps = Omit<ConditionBuilderProviderProps, 'children'>;
+
+const conditionModes = [
+  { label: 'Basic Condition', value: 'basic-condition' },
+  { label: 'Nested Condition', value: 'nested-condition' },
+  { label: 'Always True', value: 'always-true' },
+  { label: 'Always False', value: 'always-false' }
+];
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx:21]: best mark this as const (readonly)
```suggestion
const conditionModes = [
  { label: 'Basic Condition', value: 'basic-condition' },
  { label: 'Nested Condition', value: 'nested-condition' },
  { label: 'Always True', value: 'always-true' },
  { label: 'Always False', value: 'always-false' }
] as const;
```

File: packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    icon: IvyIcons.Process
+  };
+};
+
+export const ConditionBuilder = ({ onChange }: ConditionBuilderProps) => {
+  const { addConditionGroup, generateConditionString, conditionGroups, conditionMode, setConditionMode } = useConditionBuilderContext();
+
+  useEffect(() => {
+    onChange(generateConditionString());
+  }, [generateConditionString, onChange]);
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx:47]: why is this needed?

**Review by ivy-lli — COMMENTED** [review]
not finished, but here some first feedback.

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx (hunk 6 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:44]: I capitalised these objects because otherwise I got an ESLint error: https://jenkins.ivyteam.io/job/ui-components/job/add-condition-builder-to-browser/4/eslint/

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx (hunk 12 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:55]: No for some reason if i do it with Object.entries() the last array is always "parameters" no idea why, so i remove it.
grafik

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx (hunk 6 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:44]: Ahh because an export in a story is expected as a story in which you then can see in the storybook ui. 
Ok then best define this things in an additional file (like it is done in the table stories I think).
As long as you don't export it in the toplevel index, it will not be included in the bundle output

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx (hunk 13 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:54]: Or what do you think of the idea if I simply do the mapping of logicalOperatorOptions and typeOptions directly in the useConditionBuilderContext hook and directly return a corresponding array of label/value objects?

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx (hunk 12 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:55]: seems like a hack to me. we should better know why this is and fix it at the source :)

File: packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx (hunk 15 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx:47]: This is needed to ensure that the Info-Collapsible displays the current condition at all times. The info is defined outside the builder and can be edited via onChange.

File: packages/components/src/components/editor/browser/browser.stories.tsx (hunk 5 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/browser/browser.stories.tsx:152]: Oh yes, that's right 😄, I can add a flag that the condition builder is not integrated for this case.

File: packages/components/src/components/editor/condition-builder/ConditionGroup.tsx (hunk 12 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/ConditionGroup.tsx:55]: Yes I agree 😅 However, for some reason after I moved the hole Object.entries().map logic into the ConditionBuilderContext, there was no longer the issue with this parameters and i was able to remove the slice 👍

File: packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx (hunk 14 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx:21]: ah okey. could be improved maybe :) 
leave it for now :)

File: packages/components/src/components/editor/condition-builder/Condition.tsx [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+  groupIndex: number;
+  conditionsCount: number;
+}
+
+export const Condition = ({ condition, conditionIndex, groupIndex, conditionsCount }: ConditionProps) => {
+  const { updateCondition, removeCondition, typeOptions, logicalOperatorOptions, argumentInput } = useConditionBuilderContext();
+  return (
+    <Flex
+      direction='column'
+      style={{ border: 'var(--basic-border', borderRadius: 'var(--border-r2)', padding: 'var(--size-2)' }}
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/Condition.tsx:32]: move to css file

File: packages/components/src/components/editor/condition-builder/Condition.tsx [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+      className='condition-builder__condition'
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/Condition.tsx:33]: streamline with other classnames: `ui-condition-builder-condition`

File: packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx (hunk 1 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx:20]: It is not possible to generate a type of something which is not constant or already an type
But why would you need that, just make a type of the static part of the type and use it to access the value for that key.

Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx:20]: not sure too... maybe:

TypeOperator -> OperatorType (or just Operators)
ConditionType -> Operator
LogicOperator -> LogicOperators

I think I would rename the ConditionType, as this sais nothing right? it is an operator not a condition type...

File: packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+
+export type LogicOperator = {
+  'and': string;
+  'or': string;
+};
+
+export type LogicalOperator = keyof LogicOperator;
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx:27]: use one of them but not both :D
I would streamline this name with the "normal" operator type

File: packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx (hunk 15 above)
Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/ConditionBuilder.tsx:47]: maybe I'm wrong, but this looks like this will be executed a lot... At least multiple times per change in the builder.
Because when does the useEffect here know that something has changed inside?

File: packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx (hunk 18 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/ConditionBuilderContext.tsx:27]: I renamed the types no like this, what do you think?
grafik

File: packages/components/src/components/editor/condition-builder/Condition.tsx (hunk 9 above)
Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/Condition.tsx:13]: i renamed the Condition interface and the ConditionGroup interface to "ConditionData" and "ConditionGroupData". What do you think?

File: packages/components/src/components/editor/condition-builder/condition.tsx [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+  const { updateCondition, removeCondition, typeOptions, logicalOperatorOptions, argumentInput } = useConditionBuilderContext();
+  return (
+    <Flex direction='column' className={cn(conditionBorder, 'ui-condition-builder-condition')}>
+      <Flex direction='row' justifyContent='space-between'>
+        <Label>{`Condition ${conditionIndex + 1}`}</Label>
+        <Button onClick={() => removeCondition(groupIndex, conditionIndex)} icon={IvyIcons.Trash} aria-label='Remove Condition' />
+      </Flex>
+      <Flex alignItems='center' gap={2} direction='row'>
+        {argumentInput(condition.argument1, val => updateCondition(groupIndex, conditionIndex, 'argument1', val))}
+        <BasicField style={{ width: '150px' }}>
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/condition.tsx:33]: maybe move this to the css file?

File: packages/components/src/components/editor/condition-builder/condition.tsx [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+            onValueChange={val => updateCondition(groupIndex, conditionIndex, 'operator', val as Operator)}
+          />
+        </BasicField>
+        {condition.operator !== 'is true' &&
+          condition.operator !== 'is false' &&
+          condition.operator !== 'is empty' &&
+          condition.operator !== 'is not empty' &&
+          argumentInput(condition.argument2, val => updateCondition(groupIndex, conditionIndex, 'argument2', val))}
+        {conditionIndex < conditionsCount - 1 && (
+          <BasicField style={{ width: '70px' }}>
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/condition.tsx:46]: maybe move to the css file?

File: packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+      <>
+        <ConditionBuilderProvider
+          generateConditionString={generateConditionString}
+          logicalOperatorOptions={logicalOperatorOptions}
+          typeOptions={typeOptions}
+        >
+          <ConditionBuilder onChange={setCondition} />
+        </ConditionBuilderProvider>
+        <pre
+          style={{ border: 'var(--basic-border)', maxHeight: '80px', padding: 'var(--input-padding)', borderRadius: 'var(--border-r2)' }}
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.stories.tsx:33]: maybe use the known variables?
```suggestion
          style={{ border: vars.border.basic, maxHeight: '80px', padding: vars..., borderRadius: vars.border.r2' }}
```

File: packages/components/src/components/editor/condition-builder/conditionGroup.tsx [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+interface ConditionGroupProps extends Pick<ConditionProps, 'groupIndex'> {
+  group: ConditionGroupData;
+  groupCount: number;
+}
+
+export const ConditionGroup = ({ group, groupIndex, groupCount }: ConditionGroupProps) => {
+  const { updateLogicalOperator, addCondition, removeConditionGroup, conditionMode, logicalOperatorOptions } = useConditionBuilderContext();
+  return (
+    <Flex direction='column' gap={2} className='ui-condition-builder-group' data-condition-mode={conditionMode}>
+      <Flex direction='column' className={conditionMode === 'nested-condition' ? cn(conditionBorder) : ''} gap={2}>
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionGroup.tsx:24]: didn't you check what `cn` does?
```suggestion
      <Flex direction='column' className={cn(conditionBorder, conditionMode === 'nested-condition' && 'nested-condition')} gap={2}>
```

or why not add this mode as data arg? sounds more like a data arg than a classname...
```suggestion
      <Flex direction='column' className={conditionBorder} data-condition-mode={conditionMode} gap={2}>
```

File: packages/components/src/components/editor/condition-builder/conditionBuilder.tsx [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+      </ConditionBuilderProvider>
+    ),
+    applyModifier: () => {
+      return { value: value || '' };
+    },
+    infoProvider: () => <p>{value}</p>,
+    icon: IvyIcons.Process
+  };
+};
+useConditionBuilder.displayName = 'useConditionBuilder';
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.tsx:45]: why do you add a displayName to a hook? I think this makes no sense or am I wrong?
The displayName is a react thing to name components. E.g. they will be shown like that in the react dev tools.
So it is nice if the components have such a displayName, but I think this is useless on hooks.

File: packages/components/src/components/editor/condition-builder/conditionBuilder.tsx [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+    <Flex direction='column' gap={2}>
+      <BasicSelect items={conditionModes} value={conditionMode} onValueChange={val => setConditionMode(val as ConditionMode)} />
+      {conditionMode !== 'always-true' &&
+        conditionMode !== 'always-false' &&
+        conditionGroups
+          .map((group, groupIndex) => {
+            if (conditionMode === 'basic-condition' && groupIndex > 0) return null;
+            return <ConditionGroup key={groupIndex} group={group} groupIndex={groupIndex} groupCount={conditionGroups.length} />;
+          })
+          .filter(Boolean)}
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.tsx:64]: you can also filter first or not? I think then also the filter at the end would not be needed...

```suggestion
          .filter((group, groupIndex) => !(conditionMode === 'basic-condition' && groupIndex > 0))
          .map((group, groupIndex) => <ConditionGroup key={groupIndex} group={group} groupIndex={groupIndex} groupCount={conditionGroups.length} />)
```

File: packages/components/src/components/editor/condition-builder/conditionBuilderContext.tsx [hunk 25]
```diff
@@ -0,0 +1,172 @@
...
+  and: string;
+  or: string;
+};
+
+export type LogicOperator = keyof LogicOperators;
+
+export interface ConditionBuilderProviderProps {
+  typeOptions: Operators;
+  logicalOperatorOptions: LogicOperators;
+  generateConditionString: (conditionMode: ConditionMode, conditionGroups: ConditionGroupData[]) => string;
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilderContext.tsx:42]: I prefer the Array<ConditionGroupData> syntax over the ConditionGroupData[] syntax...

Comment (ivy-edp) [reply packages/components/src/components/editor/condition-builder/conditionBuilderContext.tsx:42]: So should I change all the type definitions where i put bla[] instead of Array<bla>?

Comment (ivy-lli) [reply packages/components/src/components/editor/condition-builder/conditionBuilderContext.tsx:42]: If you want. I just wanted to mention it. 
https://www.totaltypescript.com/array-types-in-typescript

File: packages/components/src/components/editor/condition-builder/conditionBuilder.css.ts [hunk 26]
```diff
@@ -1,8 +1,24 @@
...
+export const conditionBorder = style(sharedConditionStyles);
+
+export const conditionGroupBorder = style({
+  selectors: {
+    '&[data-condition-mode="nested-condition"]': sharedConditionStyles
+  }
+});
+
+export const operatorField = style({
+  width: '150px'
```
Comment (ivy-lli) [inline packages/components/src/components/editor/condition-builder/conditionBuilder.css.ts:19]: just as info, if you have px then you can also write the following if you like:
```suggestion
  width: 150
```

**Review by ivy-lli — APPROVED** [review]
Looks good to me.
Thanks for the improvements 👍
